X3: Recursive grammar: alternative container parses substitution issue

I would like to parse such text:

```
config {
	foo: 0
	bar: 1 {
		bar_child: 11
		bar_child2: 12
	}
	quux: 2
}
```

```cpp
#include <boost/spirit/home/x3.hpp>
#include <boost/fusion/include/adapt_struct.hpp>
#include <string>
#include <vector>

namespace x3 = boost::spirit::x3;

namespace ast
{
	struct config_param
	{
		std::string name;
		int value;
		std::vector<config_param> child_params;
	};

	struct config
	{
		std::vector<config_param> params;
	};
}

BOOST_FUSION_ADAPT_STRUCT(
	ast::config_param,
	name, value, child_params)

BOOST_FUSION_ADAPT_STRUCT(
	ast::config,
	params)

using comment_type = x3::rule<class comment_class>;
BOOST_SPIRIT_DECLARE(comment_type)

using identifier_type = x3::rule<class identifier_class, std::string>;
BOOST_SPIRIT_DECLARE(identifier_type)

using whitespace_type = x3::rule<class whitespace_class>;
BOOST_SPIRIT_DECLARE(whitespace_type)

using config_param_type = x3::rule<class config_param_class, ast::config_param>;
BOOST_SPIRIT_DECLARE(config_param_type)

using config_type = x3::rule<class config_class, ast::config>;
BOOST_SPIRIT_DECLARE(config_type)

using grammar_type = x3::rule<class grammar_class, ast::config>;
BOOST_SPIRIT_DECLARE(grammar_type)

using skipper_type = whitespace_type;
using iterator_type = std::string::const_iterator;
using context_type = x3::phrase_parse_context<skipper_type>::type;

const comment_type comment = "comment";
const auto comment_def = x3::lexeme['#' >> *(x3::char_ - '\n')];
BOOST_SPIRIT_DEFINE(comment)

const identifier_type identifier = "identifier";
const auto identifier_def = x3::lexeme[(x3::alpha | '_') >> *(x3::alnum | '_')];
BOOST_SPIRIT_DEFINE(identifier)

const whitespace_type whitespace = "whitespace";
const auto whitespace_def = x3::space | comment;
BOOST_SPIRIT_DEFINE(whitespace)

const config_param_type config_param = "config param";
const auto config_param_def = identifier > ':' > x3::int_ > (('{' > *config_param > '}') /* | x3::attr(std::vector<ast::config_param>()) */ );
BOOST_SPIRIT_DEFINE(config_param)

const config_type config = "config";
const auto config_def = x3::lit("config") > '{' > *config_param > '}';
BOOST_SPIRIT_DEFINE(config)

const grammar_type grammar = "code";
const auto grammar_def = config;
BOOST_SPIRIT_DEFINE(grammar)

BOOST_SPIRIT_INSTANTIATE(grammar_type, iterator_type, context_type)
```

With comment enabled, every config param must have empty `{}` after it's value. I tried to remove this requirement by providing alternative that results in the same attribute (empty vector of childs). However, if you uncomment code, it does not compile.

Is this the same bug as #434 ? Any workaround?

I haven't thoroughly looked through template backtrace but `error: invalid use of incomplete type 'boost::mpl::bind3< [...] >::type' {aka 'struct boost::spirit::x3::traits::is_substitute<std::vector<ast::config_param>, std::vector<ast::config_param>, void>'}` seemed almost ok for me as  `(('{' > *config_param > '}') | x3::attr(std::vector<ast::config_param>()))` does look like it would instantiate an alternative with exactly the same types ... incomplete type was quite misleading, didn't though it could result from failed SFINAE in all cases.

I tried to workaround it differently: 
```cpp
	struct config_param
	{
		std::string name;
		int value;
		std::optional<std::vector<config_param>> child_params;
	};

const auto config_param_def = identifier > ':' > x3::int_ > -('{' > *config_param > '}');
```
But optional vector does't compile either:
```
boost/spirit/home/x3/support/traits/container_traits.hpp:125:15: error: 'class std::optional<std::vector<ast::config_param> >' has no member named 'insert'
             c.insert(c.end(), std::move(val));
             ~~^~~~~~
```

Something causes it to treat `optional<vector>` like just `vector`

Tried example with `boost::optional<std::vector<ast::config>>` instead of `std::optional` - doesn't compile either but has a different error:

```
boost/fusion/support/is_view.hpp:31:45: error: no type named 'is_view' in 'class boost::optional<std::vector<ast::config_param> >'
                 typedef typename T::is_view type;
                                             ^~~~
```

--- a/include/boost/spirit/home/x3/support/traits/is_substitute.hpp
+++ b/include/boost/spirit/home/x3/support/traits/is_substitute.hpp
@@ -46,7 +46,7 @@ namespace boost { namespace spirit { namespace x3 { namespace traits
         {};
 
         template <typename T, typename Attribute, typename Enable = void>
-        struct is_substitute_impl : is_same<T, Attribute> {};
+        struct is_substitute_impl : mpl::false_ {};
 
         template <typename T, typename Attribute>
         struct is_substitute_impl<T, Attribute,
@@ -75,16 +75,15 @@ namespace boost { namespace spirit { namespace x3 { namespace traits
             typename enable_if<
                 is_variant<Attribute>
             >::type>
-          : mpl::or_<
-                is_same<T, Attribute>
-              , variant_has_substitute<Attribute, T>
-            >
+          : variant_has_substitute<Attribute, T>
         {};
     }
 
     template <typename T, typename Attribute, typename Enable /*= void*/>
     struct is_substitute
-        : detail::is_substitute_impl<T, Attribute> {};
+        : mpl::or_<
+              is_same<T, Attribute>,
+              detail::is_substitute_impl<T, Attribute>> {};
 
     // for reference T
     template <typename T, typename Attribute, typename Enable>

--- a/test/x3/container_support.cpp
+++ b/test/x3/container_support.cpp
@@ -24,6 +24,35 @@
 
 namespace x3 = boost::spirit::x3;
 
+
+// check if we did not break user defined specializations
+namespace check_substitute {
+template <typename T> struct foo {};
+template <typename T> struct bar { using type = T; };
+template <typename T> struct is_bar : std::false_type {};
+template <typename T> struct is_bar<bar<T>> : std::true_type {};
+}
+
+namespace boost { namespace spirit { namespace x3 { namespace traits {
+using namespace check_substitute;
+
+template <typename T, typename U>
+struct is_substitute<foo<T>, foo<U>> : is_substitute<T, U> {};
+
+template <typename T, typename U>
+struct is_substitute<T, U, std::enable_if_t<is_bar<T>::value && is_bar<U>::value>>
+  : is_substitute<typename T::type, typename U::type> {};
+}}}}
+
+namespace check_substitute {
+using x3::traits::is_substitute;
+static_assert(is_substitute<foo<int>, foo<int>>::value, "is_substitute problem");
+static_assert(!is_substitute<foo<int>, foo<long>>::value, "is_substitute problem");
+static_assert(is_substitute<bar<int>, bar<int>>::value, "is_substitute problem");
+static_assert(!is_substitute<bar<int>, bar<long>>::value, "is_substitute problem");
+}
+
+
 x3::rule<class pair_rule, std::pair<std::string,std::string>> const pair_rule("pair");
 x3::rule<class string_rule, std::string> const string_rule("string");
 

--- a/test/x3/rule3.cpp
+++ b/test/x3/rule3.cpp
@@ -7,15 +7,18 @@
 
 #include <boost/detail/lightweight_test.hpp>
 #include <boost/spirit/home/x3.hpp>
+#include <boost/fusion/include/adapt_struct.hpp>
 #include <boost/fusion/include/std_pair.hpp>
 
 #include <boost/variant.hpp>
 #include <string>
+#include <vector>
 #include <cstring>
 #include <iostream>
 #include "test.hpp"
 
 using boost::spirit::x3::_val;
+namespace x3 = boost::spirit::x3;
 
 struct f
 {
@@ -60,6 +63,25 @@ boost::spirit::x3::rule<class grammar_r, node_t> const grammar;
 auto const grammar_def = '[' >> grammar % ',' >> ']' | boost::spirit::x3::int_;
 
 BOOST_SPIRIT_DEFINE(grammar)
+
+}
+
+struct recursive_tuple
+{
+    int value;
+    std::vector<recursive_tuple> children;
+};
+BOOST_FUSION_ADAPT_STRUCT(recursive_tuple,
+    value, children)
+
+// regression test for #461
+namespace check_recursive_tuple {
+
+x3::rule<class grammar_r, recursive_tuple> const grammar;
+auto const grammar_def = x3::int_ >> ('{' >> grammar % ',' >> '}' | x3::eps);
+BOOST_SPIRIT_DEFINE(grammar)
+
+BOOST_SPIRIT_INSTANTIATE(decltype(grammar), char const*, x3::unused_type)
 
 }
 

